WordPress/gutenberg
# Global Styles: Try per block custom CSS

## What?

Adds the option to add custom CSS per block via the Styles sidebar in the Site Editor and via theme.json.
Continuation on https://github.com/WordPress/gutenberg/issues/30142, https://github.com/WordPress/gutenberg/issues/44412

## Why?

To allow more customizations.

## How?

This PR builds on the previous changes that adds the option to add custom CSS to the root via the Styles sidebar.

- Updates the global styles **context menu** to display the button that opens the Custom CSS panel for each block.
- Updates the **CustomCSSControl** to use the block specific CSS (styles.blocks.blockname.css) if a block name is defined.
-  Updates the screen, **ScreenCSS** to pass the block name to the CustomCSSCControl and to display a slightly different panel description, containing the block name (I added the block name for clarity and ease of use ).
- Outputs the per block CSS in the editors and front.

## Testing Instructions

1. Apply the PR
2. Activate the test theme, "empty theme"
3. In theme.json, add CSS to `styles.blocks.blockname.css`,
for example to the paragraph block:

```
"styles": {
		"blocks": {
			"core/paragraph": {
				"css": "color: orange;"
			}
		}
	}
```
4. View the theme and confirm that the CSS you added in theme.json is applied in the site editor, block editor, and on the front.
5. Open the Site Editor. Select "Edit" in the new navigation sidebar. Open the Styles sidebar, select Blocks.
Open the Styles sidebar, select Blocks.
6. Select the block that you added CSS to in the previous step.
7. Below the typography, colors and layout settings you should find a new panel with the text "Add your own CSS to customize the block appearance."
8. Select the button with the text "Additional block CSS" to open the CSS panel.
9. Confirm that there is a description at the top of the panel, after the back button and Heading.
Confirm that the description contains the name of the current block:
"Add your own CSS to customize the appearance of the Paragraph block."
10. Below that is an h2 heading that says "ADDITIONAL CSS" followed by 
a text area for inputting CSS.
11. Confirm that the text area shows the CSS that you added to theme.json.
12. Make a change to the existing CSS and make a new addition.
13. Confirm that the changes are applied to the blocks in the site editor.
14. Confirm that there is a help text below the text area with some instructions.
15. Next, confirm that there is an additional panel below the help text.
- The panel should be closed by default.
- The panel has an H2 heading with a button with the text "Original Theme Custom CSS"
Activate the button to open the panel.
The panel should contain a `<pre>` element with the CSS that you added to theme.json.
16. Save the style changes.
17. Go back one step to the previous panel, the panel for the block.
18. Confirm that the block preview has been updated with your styles.
19. View the front of the website and confirm that your styles are applied.


### Testing Instructions for Keyboard
The Styles sidebar in general has accessibility problems where the heading order is incorrect and where information
is not available to screen reader users.
These problems affect the sidebar as a whole and needs to be carefully tested and fixed in a separate PR.

### Testing permissions
1. Disable unfiltered HTML by adding the following constant to wp-config.php - define( 'DISALLOW_UNFILTERED_HTML', true );
2. Refresh the Site Editor, open the Styles sidebar, select Blocks. Select a block, confirm that this panel does not have a custom CSS button.

1. Apply the PR
2. Activate the test theme, "empty theme"
3. In theme.json, add CSS to `styles.blocks.blockname.css`,
for example to the paragraph block:

```
"styles": {
		"blocks": {
			"core/paragraph": {
				"css": "color: orange;"
			}
		}
	}
```
4. From the WordPress admin area, navigate to the Site Editor under Appearance, Editor.
5. Once in the Site Editor, activate the button with the aria-label "Open the editor".
6. Navigate backwards into the "Editor top toolbar". Navigate backwards past the Options and Navigation buttons, and activate the "Styles" button.
7 - Note: The styles sidebar is now opened, but it is unclear to me how to best navigate to it because activating the button does not move focus to the sidebar content.
Instead, I needed to tab forward past the navigation and option buttons, and past the editor content.
The next focusable item is then a button labeled "open Style book". This button is inside the styles panel, but it is not the button we want to access.
Instead, press the tab key to tab forwards past More styles, close, browse styles, typography, colors, layout, until you arrive at the button "Block Styles". Activate this button.
8. _Even though it is not announced that the block styles panel is opened,_ focus should now be inside the correct sidebar panel. Use tab to navigate through a list of buttons labeled with the block name. Select the button for the block that you edited in theme.json.
9. Navigate past the block options for typography, color and layout until you reach the button with the aria-label "Additional block CSS". Activate the button.
10. tab forwards to reach a text area. When the text area is in focus, a help text with instructions is also read out.
11. Confirm that there is CSS in the text area that matches what you added to theme.json.
12. Edit the CSS and make your own addition.
13. Save.
14. Reload/refresh the page and navigate back to the text area. Confirm that the updates you made have been saved.

## Screenshots or screencast 

<img width="279" alt="The button that opens the custom CSS panel, in the Styles > blocks panel" src="https://user-images.githubusercontent.com/7422055/207802193-a03dcae0-ab07-4fec-b585-3bb336c18701.png">


<img width="279" alt="The per block custom CSS panel in the site editor" src="https://user-images.githubusercontent.com/7422055/207801971-8b3419bd-aea5-478c-a9bc-7d9e19027656.png">
